Add URL Details endpoint to REST API to allow retrieval of info about a remote URL

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/18047.

Adds a new REST API endpoint to grab details from a external URL. So now if I want to get the contents of the `title` tag on `https://wordpress.org` I can send a request to this API endpoint and it will return the value in the response for use in the Gutenberg UI.

Why might this be useful? See https://github.com/WordPress/gutenberg/pull/19387.

-----

~As part of https://github.com/WordPress/gutenberg/pull/17846 we need to display the contents of the `<title>` tag from the url the user has typed into the input field.~ This has now long since shipped. Nonetheless, it might be a nice enhancement to be able to show details about the remote URL that has been selected.

This PR addresses this by adding a new **experimental** custom Gutenberg REST API endpoint. This endpoint:

* makes a request for a remote URL.
* ~parses the response in a `DOMDocument`~ we should investigate this in future (see [here](https://github.com/google/web-stories-wp/blob/a190acea43fc8dcac937d95b114d61ef4cad6aad/includes/REST_API/Link_Controller.php))
* retrieves the `title` tag contents.

To prevent abuse, the endpoint is authenticated and requires the user to have the `edit_posts` capability to be able to make a request.

## How has this been tested?

- [x] Manual testing (see below)
- [x] Automated tests

## Automated Testing Instructions

`npm run test-php` should run the unit tests which reside in `phpunit/class-wp-rest-url-details-controller-test.php`.

## Manual Testing Instructions

* Check out this PR.
* Open the Gutenberg local testing env at `http://localhost:8889/wp-admin/` and login.
* Create a new Post.
* Choose one of the following options:

### Option 1

* Open devtools and in the `console` enter:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=https://wordpress.org' } ).then( data => {
    console.log( data );
} );
```

You'll see the response returned there.

### Option 2
* Open devtools and in the `console` enter `wpApiSettings.nonce`. You should see a valid nonce returned as a string.
* Copy the nonce.
* In your browser goto: `http://localhost:8889/wp-json/__experimental/url-details/?url=https://wordpress.org&_wpnonce=%YOUR_NONCE_HERE%` - be sure to replace the nonce value with that copied from the previous step.
* You should see a valid REST API response containing the contents of the title tag from `wordpress.org` (ie: `"Blog Tool, Publishing Platform, and CMS \u2014 WordPress.org"`).



## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Scaffold out basic new endpoint

getdave committed:
Implement basic retrival of title tag from remote url

getdave committed:
Adds validation, sanitization and permissions checks.

getdave committed:
i18n fixes and docblocks

getdave committed:
Adds caching of remote request

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+					'callback'            => array( $this, 'get_title' ),
+					'args' => array(
+						'url' => array(
+							'validate_callback' => function( $param ) {
```
This could be just `'validate_callback' => 'wp_http_validate_url',`

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+							'validate_callback' => function( $param ) {
+								return $this->validate_url( $param );
+							},
+							'sanitize_callback' => function( $param ) {
```
Same as above, `'sanitize_callback' => 'esc_url_raw',`

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+		$response = null;
+
+		// Transient per URL
+		$cache_key = 'g_url_details_response_' . md5( $url );
```
you might be running into db key length issues here, not sure. `hash( 'crc32b', $url )` would help shorten that.

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+		if ( ! empty( $cached_response ) ) {
+			$response = $cached_response;
+		} else {
+			$response = wp_remote_get( $url );
```
You could simplify this bit by assigning the transient to `$response` and and only checking if that is empty to make the request. That would allow you to remove the assignment in line 146, too.

xknown replied:
I think we probably need to use `wp_safe_remote_get` instead of `wp_remote_get`, the former has some additional protections.

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+			set_transient( $cache_key, $response, HOUR_IN_SECONDS );
+		}
+
+		$body = wp_remote_retrieve_body( $response );
```
I'd do that before saving the entire response in the transient.

ocean90 replied:
The entire response shouldn't be fetched and stored at all. That's just too much unnecessary data.

Instead, I suggest to take a look at `wp_xmlrpc_server::pingback_ping()`, especially the `limit_response_size` argument for the HTTP API. To get the right data you can check some popular sites and figure out at which size the `<title>` element would be still in the response. Getting the title from the response should then be doable with a _simple_ regex.

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+
+		$required_cap = 'edit_posts';
+
+		if ( ! current_user_can( $required_cap ) ) {
```
I think it's fine to have `'edit_posts'` in line here.

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+		$html_response = $this->get_remote_url_html( $url );
+
+		if ( is_wp_error( $html_response ) ) {
+			return new WP_Error( 'no_title', __( 'Unable to retrieve title tag.', 'gutenberg' ) . $html_response->get_error_message(), array( 'status' => 404 ) );
```
If `$html_response` is a WP_Error, you could just return that here

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+
+
+
+
```
That seems like a lot of empty lines :)

obenland commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+		// Restore error level
+		libxml_use_internal_errors( $internalErrors );
+
+		return $dom;
```
Can we find a way to contain all the `DOMDocument` business in one method? Either this one or the calling one?

xknown commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,182 @@
+
+		$body = wp_remote_retrieve_body( $response );
+
+		$dom = new DOMDocument( '1.0', 'UTF-8' );
```
We probably need to make sure the `DOMDocument` class exists, otherwise there might be PHP fatals if the `php-xml/php-dom` extensions are not enabled.

obenland committed:
Update with feedback

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,131 @@
+		$cache_key = 'g_url_details_response_' . hash( 'crc32b', $url );
+
+		// Attempt to retrieve cached response.
+		$title = null;//get_transient( $cache_key );
```
Whoops! We've lost our cache. Can we use `WP_DEBUG` here to avoid this in future?

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,131 @@
+		$title = null;//get_transient( $cache_key );
+
+		if ( empty( $title ) ) {
+			$request                = wp_safe_remote_get( $url, array(
```
We now have a single function which does

1. Remote URL fetching
2. Parsing of a `title` tag.
3. Caching.

If we consider that we may also need to fetch other types of details from a site (eg: favicon) would it be better to separate the concerns here? That way we can call "get site" as an independent method and then parse the info we need using the requisite extractor method.

obenland replied:
Sure, let's refactor when the time comes.

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,131 @@
+				return new WP_Error( 'no_response', __( 'The source URL does not exist.', 'gutenberg' ), array( 'status' => 404 ) );
+			}
+
+			// Work around bug in strip_tags():
```
Which bug is this? Would be great to have something to reference here.

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,131 @@
+			// Work around bug in strip_tags():
+			$remote_source = str_replace( '<!DOC', '<DOC', $remote_source );
+			$remote_source = preg_replace( '/[\r\n\t ]+/', ' ', $remote_source ); // Normalize spaces.
+			$remote_source = preg_replace( '/<\/*(h1|h2|h3|h4|h5|h6|p|th|td|li|dt|dd|pre|caption|input|textarea|button|body)[^>]*>/', "\n\n", $remote_source );
```
What is this doing?

getdave reviewed (commented):
@obenland Thanks for the changes! Really appreciate you taking this on and making it much better.

Left some small comments. Interested in what you think.

We need to wrangle some more feedback from the [REST API group in Core](https://wordpress.slack.com/messages/C02RQC26G).  

We also need a rebase to make the playground pass (it's happened on all PRs recently).

Lastly, what do you think about [this comment on the Issue](https://github.com/WordPress/gutenberg/issues/18047#issuecomment-546040014)?

obenland committed:
Tie up loose ends

obenland committed:
Remove unneed replacements

TimothyBJacobs commented:
Has @mnelson4's [comment](https://github.com/WordPress/gutenberg/issues/18047#issuecomment-546040014) been addressed?

getdave commented:
@kadamwhite We'd really appreciate your expertise and input on this if at all possible.

getdave commented:
@obenland Shall we try and move this forward in our next 20% time? What should we do to prepare for that so we can make progress at that time?

obenland commented:
Sure. I think double-checking https://github.com/WordPress/gutenberg/pull/18042#issuecomment-552009258 would be the next step

getdave commented:
> Sure. I think double-checking [#18042 (comment)](https://github.com/WordPress/gutenberg/pull/18042#issuecomment-552009258) would be the next step

I've [addressed the issue raised here](https://github.com/WordPress/gutenberg/issues/18047#issuecomment-570160408).

getdave commented:
Waiting on further feedback here. The outstanding issue blocking this from merging is the concern about exposing data about websites behind a firewall.

mnelson4 commented:
From what I could test, this doesn't expose data local data. Although I don't have a network behind a firewall etc, so I can't really comment about that.

I tested this out and it worked fine. It did seem a bit odd the response wasn't actually JSON, just a string of the site's title. (I suppose returning JSON leaves the door open to returning other metadata about the site.) I wasn't aware of anywhere else in the REST API that just does that, but it's not the end of the world.

aduth commented:
My comment at https://github.com/WordPress/gutenberg/issues/18061#issuecomment-581484853 highlights another potential use-case for an endpoint like this, where the title preview shown in the preview state of `LinkControl` may be something we would want to retrieve using this endpoint.

![LinkControl preview state](https://user-images.githubusercontent.com/1779930/73668274-a57f7b00-4673-11ea-8624-1e39b708b381.png)

One thought that occurs to me is: If the given URL can be determined to be from the current site, would we still want to use the same scraping procedure, or could it be shortcut since it might be simpler to evaluate as known to the current site?

Pseudo-code:

```php
$post_id = url_to_postid( $url );
if ( ! empty( $post_id ) ) {
    return get_the_title( $post_id );
}
```

- https://developer.wordpress.org/reference/functions/url_to_postid/
- https://developer.wordpress.org/reference/functions/get_the_title/

getdave commented:
> I tested this out and it worked fine. It did seem a bit odd the response wasn't actually JSON, just a string

@obenland Should we change this to return JSON?

getdave commented:
> My comment at #18061 (comment) highlights another potential use-case for an endpoint like this, where the title preview shown in the preview state of LinkControl may be something we would want to retrieve using this endpoint.

@aduth Yes agreed. My PR handles `title` only but this and favicon and what you suggest are good ideas.

getdave commented:
We should check this handler strips off any query params from the URL as these could be used to manipulate a remote website via the URL.

aduth commented:
> We should check this handler strips off any query params from the URL as these could be used to manipulate a remote website via the URL.

I'm not sure I understand what's meant by this? Or, at least, I imagine removing query parameters could produce incorrect details for many URLs, such as WordPress sites using the [default permalink structure](https://wordpress.org/support/article/using-permalinks/#default-ugly). While it's true that a poorly-implemented page could be susceptible to [CSRF via GET](https://en.wikipedia.org/wiki/Cross-site_request_forgery#HTTP_verbs_and_CSRF), I'm not sure it's something we ought to be held accountable for:

>Request methods are considered "safe" if their defined semantics are essentially read-only; i.e., the client does not request, and does not expect, any state change on the origin server as a result of applying a safe method to a target resource. [...] This definition of safe methods does not prevent an implementation from including behavior that is potentially harmful, that is not entirely read-only, or that causes side effects while invoking a safe method. What is important, however, is that **the client did not request that additional behavior and cannot be held accountable for it.** [...] Of the request methods defined by this specification, the GET, HEAD, OPTIONS, and TRACE methods are defined to be safe.

https://tools.ietf.org/html/rfc7231#section-4.2.1

getdave commented:
> I'm not sure I understand what's meant by this? Or, at least, I imagine removing query parameters could produce incorrect details for many URLs, such as WordPress sites using the default permalink structure. While it's true that a poorly-implemented page could be susceptible to CSRF via GET, I'm not sure it's something we ought to be held accountable for:

Agreed. I wrote this when I asked my friend who is a professional penetration tester what other things we could do to avoid the endpoint being manipulated. But you are right that this approach wouldn't work in our context and we can't be responsible for everything.

draganescu reviewed (approved):
I have tested this and works very well. Considering the security testing and discussion previously on this PR I will approve this since it does not introduce any unseen or un-testable behavior.

getdave commented:
Next steps here is to get this working within `LinkControl` so that it shows the title of the remote URL in the search results.

Also might be worth investigating whether we can do something similar for Favioncs and store them as datauris.

getdave commented:
Looks like @draganescu approved this quite a while ago and I forgot/missed it. @youknowriad do you think you would you be able to take a look at this and confirm you'd still be happy for this to land?

youknowriad commented:
Seems like something useful to me but I'd like to defer to the REST API team and @TimothyBJacobs as I know they were discussing experimental Endpoints;

Also it would be good to clarify some things:

 - Why it's experimental?
 - Is this supposed to land in 5.5? or Is it required for a feature that will land in WP 5.5? if not, we'd have more room to land in the plugin as experimental or not before backporting to Core. If it's supposed to be 5.5, a Core ticket might be needed.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,122 @@
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function get_remote_url_permissions_check() {
+		if ( ! current_user_can( 'edit_posts' ) ) {
```
This breaks things for custom post types. I'd suggest taking a look at the permissions check in the Themes controller.

getdave replied:
I look at [the Themes Controller here](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L60) and updated the code to match.

TimothyBJacobs commented:
Left a comment about permission handling, after that I'm fine with this merging into GB.

> Is this supposed to land in 5.5? or Is it required for a feature that will land in WP 5.5? If it's supposed to be 5.5, a Core ticket might be needed.

Yep a trac ticket would be needed for this.

ocean90 commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,122 @@
+	 */
+	private function get_remote_url_title( $url ) {
+		// Transient per URL.
+		$cache_key = 'g_url_details_response_' . hash( 'crc32b', $url );
```
Why not `md5()`? Also, `g_remote_url_title` might be a better name now.

obenland replied:
md5 would create a transient key that would bump up to or exceed [the 64 char limit of `option_name`](https://core.trac.wordpress.org/ticket/15058).

ocean90 replied:
@obenland 64 was the limit 5 years ago, [it's now 191](https://core.trac.wordpress.org/changeset/34030). Though, the transient name itself can only have 172 or less characters due to the `_transient_` prefixes.

getdave replied:
@obenland @ocean90 Could you advise whether this is a blocker? Thanks.

swissspidy replied:
Switchinh to md5 should be easily possible

getdave replied:
Done

spacedmonkey commented:
We have written something very very similar for the web-stories project. If you want take anything from it, do feel free - https://github.com/google/web-stories-wp/blob/a190acea43fc8dcac937d95b114d61ef4cad6aad/includes/REST_API/Link_Controller.php

spacedmonkey commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,122 @@
+			$request                = wp_safe_remote_get( $url, array(
+				'timeout'             => 10,
+				'limit_response_size' => 153600, // 150 KB.
+			) );
```
You should check the response is not wp error here and that response was a 200.

getdave replied:
Fixed.

spacedmonkey commented on lib/rest-api.php:
```diff
@@ -54,6 +54,17 @@ function gutenberg_filter_oembed_result( $response, $handler, $request ) {
+/**
+ * Registers the REST API routes for URL Details.
+ *
+ * @since 5.0.0
```
Change version number.

spacedmonkey commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,122 @@
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/title',
```
What if want other fields in the future. Why not just have a details endpoint, with different fields?

swissspidy replied:
Agreed. This should be more future proof and anticipate more fields than just the titlee.
Devs would want to add additional fields, for example to also return the URL's favicon.

getdave replied:
I've updated to allow for this.

draganescu commented:
@getdave since the screen and the block will not be included in 5.5 neither is this endpoint required to land. On the other hand this is a very handy endpoint to have when building the navigation. @spacedmonkey left some comments, do you think you can pick this PR back up?

getdave commented:
> @getdave since the screen and the block will not be included in 5.5 neither is this endpoint required to land. On the other hand this is a very handy endpoint to have when building the navigation. @spacedmonkey left some comments, do you think you can pick this PR back up?

Missed this! I'd _like_ to pick it up, whether I will have time is another matter 😄

getdave committed:
Account for Custom Post Types in permissions check

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r441597385

swissspidy commented:
@getdave I might be able to help. There are some use cases that I‘d love to be able to support with such an endpoint.

getdave committed:
Improve error handling

getdave committed:
Improve args checking

getdave committed:
Refactor to enable query of more data in future.

getdave committed:
Use md5

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r441624118

getdave committed:
Extract utility functions

getdave commented:
> * Why it's experimental?

@youknowriad I assumed such things _had_ to be experimental? Are you or perhaps @TimothyBJacobs able to advise on whether it would be ok to merge this _without_ the experimental prefix?

getdave commented:
Now writing some unit tests for this.

getdave committed:
Fix lint errors

getdave committed:
Add unit test scaffold

getdave committed:
Attempt to mock HTTP requests for testing. Needs work.

getdave committed:
Use DIR convention

getdave commented:
Status: unit tests need lots of work. Mainly around successfully mocking out the HTTP requests so we can have robust repeatable tests.

TimothyBJacobs commented:
The thing to keep in mind, is if it isn't marked as experimental, then we need to maintain BC for it. So generally REST endpoints are merged into GB as experimental, and then stabilized shortly before they are merged to Core.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,201 @@
+		$data = $this->get_cache( $cache_key );
+
+		// Return cache if valid data.
+		if ( false && ! is_wp_error( $data ) && ! empty( $data ) ) {
```
```suggestion
		if ( ! empty( $data ) ) {
```

* stray `false` (this condition is never met)
* errors are never cached (and shouldn't be!)

getdave replied:
> stray false (this condition is never met)

About to replace this with an `apply_filters()` call.

getdave replied:
> errors are never cached (and shouldn't be!)

Guess I'm being overly cautious.

getdave committed:
Allow opt out of cache functionality via filter

getdave committed:
Force tests to opt out of cache. Fix broken tests.

getdave committed:
Adjust 404 case to use standard response code

getdave committed:
Add dedicated methods to mock success/failure HTTP responses

getdave committed:
Add test for empty body in response.

This should trigger a failed response because we can’t read details from a URL which doesn’t provide a HTTP body.

getdave committed:
Add ability to filter request args. Update tests to account.

getdave commented:
Next up - write a schema for the endpoint.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,213 @@
+		$data = $this->get_cache( $cache_key );
+
+		// Return cache if valid data.
+		if ( apply_filters( 'rest_url_details_use_cache', true ) && ! is_wp_error( $data ) && ! empty( $data ) ) {
```
This needs filter docs.

getdave replied:
Added

TimothyBJacobs commented:
This looks good to me for an initial go. I do think that the schema should be added before it is merged though.

getdave commented:
> schema should be added before it is merged though.

Yep that's all good. I'll get that sorted.

getdave committed:
Document cache filter

getdave committed:
Iniital attempt at adding schema

getdave committed:
Add test for unautenticated user.

getdave committed:
Adjust schema to use WP convention

getdave committed:
Adjust schema defaults

getdave committed:
Add endpoint args schema test

beaulebens commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,246 @@
+	 * @return string the title tag contents (maybe empty).
+	 */
+	private function get_title( $html ) {
+		preg_match( '|<title>([^<]*?)</title>|is', $html, $match_title );
```
Might need to allow for arbitrary attributes on the `<title>`?

getdave replied:
That sounds like a good plan. Here's a first pass:

```
<title( [a-z]+="[\w\d\s]+")*>([^<]*?)<\/title>
```

Ultimately we're probably going to use `DOMDocument` to better "query" the HTML, but I'd like to get this "right".

beaulebens replied:
It gets kind of painful trying to match all possible (valid, let alone "allowed") strings here, this SO post tries to get there -- https://stackoverflow.com/questions/317053/regular-expression-for-extracting-tag-attributes

Maybe we have something somewhere else in WP code that we can re-use?

> Ultimately we're probably going to use DOMDocument to better "query" the HTML, but I'd like to get this "right".
👍

getdave replied:
Maybe https://regex101.com/r/U9Yqqg/2.

The quicker we follow up with something more robust the better!

draganescu replied:
Hi! What kind of extra title attributes would this use?

getdave replied:
It is an edge case for sure.

We could just roll with not supporting attrs given we know I will upgrade to a more robust method of interogating the HTML in a follow up.

beaulebens replied:
> Hi! What kind of extra title attributes would this use?

FWIW my suggestion was just to make sure that we allow for (and ignore) them, not that we actually need to use them for anything. I don't know that they are very common, but they are [technically valid](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/title#Attributes).

draganescu replied:
> We could just roll with not supporting attrs given we know I will upgrade to a more robust method of interogating the HTML in a follow up.

Yes to this imho! Thanks @beaulebens for the clarification!

getdave replied:
Raised https://github.com/WordPress/gutenberg/issues/27762 to followup.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -77,7 +77,17 @@ public function parse_url_details( $request ) {
+		 * @param string $url the attempted URL.
+		 * @param string $cache_key the generated cache key for the given URL.
+		 */
+		if ( apply_filters( 'rest_url_details_use_cache', $use_cache = true, $url, $cache_key ) && ! is_wp_error( $data ) && ! empty( $data ) ) {
```
```suggestion
		if ( apply_filters( 'rest_url_details_use_cache', true, $url, $cache_key ) && ! is_wp_error( $data ) && ! empty( $data ) ) {
```

what's the use case for this filter though? Also, still would not check for `WP_Error` instances, as it leads to false assumptions. If anything I'd use `is_array()`

getdave replied:
> what's the use case for this filter though?

If I'm completely honest I added the filter as a way to aid with testing. I acknowledge that isn't a good reason to alter the system under test but it does make things a lot simpler. Also I feel there will be scenarios (yet to be identified) where folks don't want to use the cache under certain constraints. With the filter, they can check for specific `$url`s and opt in/out of cache as required.

Thinking about this again however, I could simply use [the existing filters in `get_transient()`](https://core.trac.wordpress.org/browser/tags/5.6/src/wp-includes/option.php#L825) within the tests to force return an empty value. That would remove the need for this `rest_url_details_use_cache` filter if we feel it is unnecessary. 

> Also, still would not check for WP_Error instances, as it leads to false assumptions. If anything I'd use is_array()

The argument about leading to false assumptions is a good one. I will update to `is_array()` as you suggest here.

getdave replied:
I've removed the filter and utilised the built in mechanism in transients to return a known value to avoid using the cache during the tests.

I have however, added a new tests to tests that if the cache value exists the the code will actually use it. Again we use the same mechanic to do this.

Let me know what you think and whether you'd be happy to 👍 this now.

getdave commented:
> I do think that the schema should be added before it is merged though.

@TimothyBJacobs Might you be able to advise on the schema? I'm new to this and I just want to make sure I've understood and implemented correctly (including associated tests). Your expertise here would be very much appreciated.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,246 @@
+				'title' => array(
+					'description' => __( 'The contents of the <title> tag from the URL.', 'gutenberg' ),
+					'type'        => 'string',
+					'default'     => '',
```
We don't need to include a `default` here.

getdave replied:
Fixed.

TimothyBJacobs reviewed (commented):
Schema looks good! Just a minor change.

draganescu commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,246 @@
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'timeout'             => 10,
```
Why isn't the default of 5 enough?

getdave replied:
I actually dont know. 5 should be plenty right? If it takes longer we dont want to wait around.

draganescu replied:
Yes I think so too!

getdave replied:
Fixed

getdave committed:
Removes filter to bypass cache.

This already exists in core as https://core.trac.wordpress.org/browser/tags/5.6/src/wp-includes/option.php#L785

getdave committed:
Use existing transient filter to test cache and bypass for majority of tests

Fixes https://github.com/WordPress/gutenberg/pull/18042#discussion_r543195783

getdave committed:
Use default timeout.

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r543598977

getdave committed:
Removes unnecessary default from schema

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r543541341

getdave commented:
Looking to land this pre-Xmas. Anyone around to 👍 ?

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,233 @@
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( current_user_can( 'edit_posts' ) ) {
```
Isn't that already covered by the loop below?

getdave replied:
This is based on the Themes controller. 

See also https://github.com/WordPress/gutenberg/pull/18042#discussion_r441597385

getdave replied:
@swissspidy I think this is probably ok given what @TimothyBJacobs [previously indicated](https://github.com/WordPress/gutenberg/pull/18042#discussion_r441597385).

TimothyBJacobs replied:
Yeah this is a necessity. Would probably be nice to introduce a meta cap in Core at some point...

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,233 @@
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'limit_response_size' => 153600, // 150 KB.
```
```suggestion
			'limit_response_size' => 150 * KB_IN_BYTES,
```

getdave replied:
Updated.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,233 @@
+			return $response_body;
+		}
+
+		$data = array(
+			'title' => $this->get_title( $response_body ),
+		);
```
Should the data be filterable here so that anyone can add more data to it if they want to?

getdave replied:
I agree this would be useful. It will make the endpoint extensible so that folks can parse out whatever they need. We'll probably want to cater for things such as favicon in the future by default but this opens the door for devs to retrieve additional info if they need.

Also I added a test to cover this.

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,233 @@
+		if ( ! is_array( $data ) ) {
+			return;
+		}
+		return set_transient( $key, wp_json_encode( $data ), HOUR_IN_SECONDS );
```
The TTL should perhaps be filterable here?

getdave replied:
I've made this filterable.

getdave commented:
I'll be picking this up again this coming week.

getdave committed:
Use predefine constants for response size

Resolves https://github.com/WordPress/gutenberg/pull/18042#discussion_r545265373

getdave committed:
Rearrange test code order

getdave committed:
Adds ability to filter cache expiration

Addresses https://github.com/WordPress/gutenberg/pull/18042#pullrequestreview-554823533

getdave committed:
Allow filtering the data retrived for a given URL

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r545266084

getdave commented:
@swissspidy Happy New Year to you.

I think this PR should now be ready for merge. What do you think? 

I'm aware of more functionality I'd like to add, but I'm keen to merge early and iterate if possible.

TimothyBJacobs commented:
If we want to make it possible for plugins to customize the endpoint data, we should use the existing APIs and patterns for other REST API controllers. In this case, that is using the register fields API (`add_additional_fields_schema` and `add_additional_fields_to_object`) and filtering on the entire response object. I'd take a look at the posts controller in Core for an example.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,260 @@
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'no_response', get_status_header_desc( 404 ), array( 'status' => WP_Http::NOT_FOUND ) );
```
`get_status_header_desc` is used for the HTTP status code description and isn't translated. We should use a specific error message of our own here.

getdave replied:
Updated

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,260 @@
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @access public
```
`@access` can be dropped.

getdave replied:
Done

getdave commented:
@TimothyBJacobs Thanks for the review. 

> If we want to make it possible for plugins to customize the endpoint data, we should use the existing APIs and patterns for other REST API controllers. 

This sounds good. Happy to learn about this.

> ...that is using the register fields API 

Ok, I'm not familiar enough with this. Will need to do some learning.

> `add_additional_fields_schema` 

I looked up [the docs for this](https://developer.wordpress.org/reference/classes/wp_rest_controller/add_additional_fields_schema/). They are a little...sparse 😄 .

Looking [at Posts Controller](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2077-L2079), I'm thinking it's a case of calling `add_additional_fields_schema()` [here](https://github.com/WordPress/gutenberg/pull/18042/files#diff-88b1c5b30a440fafa8745f5eaeff26e63d5a372bc460895eb9c59036b94cac11R63) passing the schema assigned to the instance. 

> and `add_additional_fields_to_object`) 

As above I took a look at [Posts Controller](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1842).

My understanding is that this is a standard method of adding properties to a `$data` response object? 

> and filtering on the entire response object. 

So is that adding [a filter like this one](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1872)?

> I'd take a look at the posts controller in Core for an example.

Yup 👍 

That said, are there any more in-depth docs on this process and how/why it is being used? I feel like I'm doing a bit of copy/paste here but potentially missing context as to why I'm doing it and what the benefits are. 

Any info much appreciated.

getdave commented:
> @TimothyBJacobs Thanks for the review.
> 
> > If we want to make it possible for plugins to customize the endpoint data, we should use the existing APIs and patterns for other REST API controllers.
> 
> This sounds good. Happy to learn about this.
> 
> > ...that is using the register fields API
> 
> Ok, I'm not familiar enough with this. Will need to do some learning.
> 
> > `add_additional_fields_schema`
> 
> I looked up [the docs for this](https://developer.wordpress.org/reference/classes/wp_rest_controller/add_additional_fields_schema/). They are a little...sparse 😄 .
> 
> Looking [at Posts Controller](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2077-L2079), I'm thinking it's a case of calling `add_additional_fields_schema()` [here](https://github.com/WordPress/gutenberg/pull/18042/files#diff-88b1c5b30a440fafa8745f5eaeff26e63d5a372bc460895eb9c59036b94cac11R63) passing the schema assigned to the instance.
> 
> > and `add_additional_fields_to_object`)
> 
> As above I took a look at [Posts Controller](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1842).
> 
> My understanding is that this is a standard method of adding properties to a `$data` response object?
> 
> > and filtering on the entire response object.
> 
> So is that adding [a filter like this one](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1872)?
> 
> > I'd take a look at the posts controller in Core for an example.
> 
> Yup 👍
> 
> That said, are there any more in-depth docs on this process and how/why it is being used? I feel like I'm doing a bit of copy/paste here but potentially missing context as to why I'm doing it and what the benefits are.
> 
> Any info much appreciated.

As discussed in DM on WP.org Slack I'm on the right track with my comments above. I'll review https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/#class-inheritance-wp_rest_controller and move forward from there.

getdave committed:
Remove @access comments

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r551440630

getdave committed:
Prefer custom error message over get_status_header_desc

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r551440380

getdave committed:
Utilise add_additional_fields_schema

Addresses https://github.com/WordPress/gutenberg/pull/18042#issuecomment-754088796

getdave committed:
Utilise add_additional_fields_to_object

Addresses https://github.com/WordPress/gutenberg/pull/18042#issuecomment-754088796

getdave committed:
Enable filtering of response object

Filter on both cached and uncached response.

getdave committed:
Add test to assert on ability to filter uncached responses

getdave committed:
Rename url placeholder literal

getdave committed:
Ensure accurately testing for correctly passed cached variable in filter args

getdave committed:
Update to test filtering of both cached and uncached responses

getdave commented:
@TimothyBJacobs I've updated as per your comments.

I've added tests to cover the ability to filter both cached and uncached response types.

Is this now ready for merge under `__experimental`?

Feel free to note down any additional enhancements you don't see as blockers to initial merge but would like to see in future. I'll create Issues and address once merged.

getdave committed:
Fixing linting

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+	 */
+	public function parse_url_details( $request ) {
+
+		$url = untrailingslashit( $request->get_param( 'url' ) );
```
```suggestion
		$url = untrailingslashit( $request['url'] );
```

The array offset syntax is preferred in Core.

getdave replied:
Done

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+		 * @param array $response_body The response body (HTML) for the given URL.
+		 * @param string $url The attempted URL.
+		 */
+		$data = apply_filters(
```
This filter shouldn't exist anymore. The general response filter can be used.

getdave replied:
I think it might still be required as this is 

1. Filtering the data that is going into cache.
2. Filtering directly on the response from the remote URL rather than the full response being returned.
3. A lot easier to interact with than a `WP_Rest_Response` object.

swissspidy replied:
@TimothyBJacobs How else would someone add additional fields to the response while having access to the `$response_body`?

For example, I'd want to add an `image` field which grabs an image from the page's source.

TimothyBJacobs replied:
Ah, my mistake I didn't notice the http response was being added to the "prepare" filter. IMO we should make that available there after `$url`.

getdave replied:
@TimothyBJacobs Thanks for your suggestion. I just want to clarify we're on the same lines here:

Initially, I'm not _sure_ (I could be wrong!) what you suggest above will address the issues I raised in https://github.com/WordPress/gutenberg/pull/18042#discussion_r555842355.

The `rest_url_details_url_data` filter allows a developer to _easily_ interrogate the HTTP response from the remote URL and prepare _additional_ items in the `$data` array.

For example:

```php
add_filter(
    'rest_url_details_url_data',
    function($data, $response_body) {
        // parse_og_title is a function I made up for illustration purposes!
        return $data['og_title'] = parse_og_title( $response_body );
    }
);
```

If we add the remote URL HTTP `$response_body` as a argument to the `rest_prepare_url_details` filter like this:

```php
// $response_body is the HTTP response from the remote URL (eg: HTML).
return apply_filters( 'rest_prepare_url_details', $response, $url, $request, $response_body );
```

...then it's going to make it _hard_ for the developer to achieve the same thing because they will have to modify the `WP_REST_Response` object to insert the new `$data` there. Indeed, [I can't see any methods on that class](https://developer.wordpress.org/reference/classes/wp_rest_response/) that allow us to do that.

For example:

```php
add_filter(
    'rest_prepare_url_details',
    function(WP_REST_Response $response, $url, $request, $response_body) {
        $additional_data = array(
            'og_title' => parse_og_title( $response_body )
        );

        // Now I have the additional data how do I append it to the $response?????
    }
);
```

@TimothyBJacobs I have two questions

1. Can you advise on how we could achieve the goal of allow developers to parse additional data out of the remote URL HTTP response to include in the rest response using the `rest_prepare_url_details` filter only (as you propose)?
2. Do you feel the additional complexity (subjective) posed by doing this via `rest_prepare_url_details` is worth it?

I appreciate your time here. I feel like we're nearly there but I do want to get this last piece right because I believe it's an important part of the puzzle in allowing this endpoint to be truly flexible. 🙇

TimothyBJacobs replied:
Sure!

They can do `$response->data['my_custom_data'] = "value"`;

I'd just switch the order of `$request, $response_body` in the filter.

getdave replied:
Ah `$response->data`, like I've been doing the whole time in the unit tests! Whoops. 

Right, that's clear now. Let me sort it. Thanks!

getdave replied:
@TimothyBJacobs As discussed in a DM on Slack, after looking again I've realised the approach suggested will not wor. This is because we do not cache the **full response from the remote URL**. Rather we cache what has been parsed from the response.

When returning the value from _cache_, the `rest_prepare_url_details` filter does not have access to the response from the remote URL. Therefore we cannot use the `rest_prepare_url_details` to parse extra data from the remote URL.

Options:

1. Don't use a cache and pass the remote URL response to the `rest_prepare_url_details` filter.
2. Retain use of a cache and continue to use x2 filters (as currently in the code of this PR).

I'm inclined towards `#2`. Could we go with that in order to get this PR merged. It is after all experimental and can be adjusted in further followups if necessary.

Perhaps you could let me know? Many thanks

TimothyBJacobs replied:
My concern with the double filtering in the REST controllers is that we're introducing an atypical pattern for modifying a core REST API response. This is further complicated by the fact that the "traditional" filter isn't particularly helpful without the URL response.

IMO it is important that the cache be as transparent to plugin developers as possible. By storing the parsed URL details in the cache, I think we break this transparency. If there is an existing cache hit, a developer won't see their filtered data until the cache expires. This is more problematic if they are utilizing additional request parameters to customize the return value since the parsed URL details may have been cached from a request that didn't have those request parameters.

I think not caching the response is fairly safe since this is a privileged endpoint and is done in response to a fairly explicit user action. However, if we don't want to go that route, IMO we should just cache the raw HTTP response. That would allow for utilizing the single filter that developers would traditionally reach for, and should result in a transparent caching layer.

getdave replied:
> IMO we should just cache the raw HTTP response

I've updated to implement this approach.

> That would allow for utilizing the single filter that developers would traditionally reach for, and should result in a transparent caching layer.

Done.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+		 * @param bool             $from_cache Whether the response is from the cache.
+		 * @param WP_REST_Request  $request  Request object.
+		 */
+		return apply_filters( 'rest_url_details_prepare_response', $response, $url, $from_cache, $request );
```
```suggestion
		return apply_filters( 'rest_prepare_url_details', $response, $url, $from_cache, $request );
```

Also, is the `from_cache` parameter necessary? I don't think that should be something plugins rely on.

getdave replied:
I thought it might be useful to have the ability to distinguish between response from cache or not. No other reason. Would that _not_ be useful?

TimothyBJacobs replied:
I think the caching should be considered an implementation detail.

getdave replied:
TBH we don't have any use case for needing the `from_cache` identifier so let's reduce the complexity and not overengineer. I'll remove it.

getdave replied:
Removed `$from_cache` and updated filter.

getdave replied:
Done

swissspidy replied:
I still see `$from_cache` here

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+		}
+
+		return new WP_Error(
+			'rest_user_cannot_view',
```
```suggestion
			'rest_cannot_view_url_details',
```

getdave replied:
Done

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+		 * Can be used to adjust the time until expiration in seconds for the cache
+		 * of the data retrieved for the given URL.
+		 *
+		 * @param int HOUR_IN_SECONDS the time until cache expiration in seconds.
```
```suggestion
		 * @param int $ttl The time until cache expiration in seconds.
```

This needs to be a variable name.

getdave replied:
Done

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+			$from_cache = true;
+			$response = rest_ensure_response( json_decode( $cached_response, true ) );
+
+			/**
```
When a filter appears multiple times, only the primary version should be fully documented. The other instances should be documented using `/** This filter is documented in path/to/file.php */`

getdave replied:
I didn't know that. Will do.

getdave replied:
Done.

TimothyBJacobs commented:
Looking good thanks @getdave! I've left some bits of feedback, I think we're almost there!

getdave commented:
> Looking good thanks @getdave! I've left some bits of feedback, I think we're almost there!

Thanks. I'll get that sorted asap!

getdave committed:
Use array offset syntax prefered by Core

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r555790183

getdave committed:
Remove from cache and update filter

See https://github.com/WordPress/gutenberg/pull/18042#discussion_r555792872

getdave committed:
Update filter name to align with standards

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r555792872

getdave committed:
Update response code to align with Core standards

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r555793039

getdave committed:
Update cache ttl in docs to use variable

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r555794092

getdave committed:
Use comment to reference canonical version of filter docs

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r555795109

getdave commented:
> > Looking good thanks @getdave! I've left some bits of feedback, I think we're almost there!
> 
> Thanks. I'll get that sorted asap!

Ok @TimothyBJacobs I think that's all addressed now apart from [your suggestion to remove the additional filter](https://github.com/WordPress/gutenberg/pull/18042#discussion_r555791516) which I'd appreciate some further input on if that's ok?

swissspidy commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -0,0 +1,291 @@
+			 * @param bool             $from_cache Whether the response is from the cache.
+			 * @param WP_REST_Request  $request  Request object.
+			 */
+			return apply_filters( 'rest_url_details_prepare_response', $response, $url, $from_cache, $request );
```
Wasn't there agreement on removing `$from_cache` here?

getdave committed:
Update to cache entire HTTP response body and remove additional filter

Addresses https://github.com/WordPress/gutenberg/pull/18042#discussion_r563470920

getdave committed:
Tweak code comments

getdave commented:
@TimothyBJacobs I've updated the endpoint to use a single filter and a "transparent cache" as we discussed above. This should now be good to go.

Looks like we have some automated test issues which appear to be a bug with one of the packages rather than anything wrong with this PR. I'll rebase this PR again early next week to see if we can go all green then.

If you could take a look at the changes and advise that would be great.

TimothyBJacobs commented:
Looks good to me! Though I'm not sure why we need to `json_encode` and `json_decode` the cache. It is just a string, right?

getdave committed:
Remove need to encode/decode cache value

getdave commented:
> Looks good to me! Though I'm not sure why we need to `json_encode` and `json_decode` the cache. It is just a string, right?

You are correct. Updated.

TimothyBJacobs reviewed (approved):
PHP Unit tests are failing, but they are also failing on `master`.

swissspidy commented:
> PHP Unit tests are failing, but they are also failing on `master`.

See https://wordpress.slack.com/archives/C02QB2JS7/p1611916183312300

getdave commented:
Rebase - I will merge this as soon as it's 🟢 . I'm keeping an eye.

getdave commented:
**Update** - failing e2e tests are not a product of this PR. I'm going to rebase again tomorrow in the hope that someone will have resolved them.

I'll merge this asap.

getdave commented:
Merged 🎉 

I will raise followups to add extra functionality.

Thanks to everyone for your help in getting this over the line (particularly @TimothyBJacobs and @swissspidy).

obenland commented:
Congrats @getdave! Nice work

getdave commented:
@TimothyBJacobs At some point if this endpoint is accepted would I need to follow up with this in WordPess Core?

TimothyBJacobs commented:
Whenever the JS feature that uses this feature is slated to be included in Core, we'll create a Core ticket for merging the URL details endpoint. I can take care of merging it into Core.

getdave commented:
> Whenever the JS feature that uses this feature is slated to be included in Core, we'll create a Core ticket for merging the URL details endpoint. I can take care of merging it into Core.

That's great thanks. Any chance I could get involved there? Only Ive not committed to WP Core and this seems like a good opportunity.

TimothyBJacobs commented:
> That's great thanks. Any chance I could get involved there? Only Ive not committed to WP Core and this seems like a good opportunity.

Yeah, you can absolutely take the lead on that!

getdave commented:
Here is a follow up https://github.com/WordPress/gutenberg/pull/28791
